### ISSUE TITLE
My AsyncInterceptor which injects an Async Logger is intercepting the Logger requests causing a loop

### ISSUE THREAD

**Issue by assadfaizi** [JSkimming/Castle.Core.AsyncInterceptor#49]
Hi,

I've managed to Implement your IAsyncInterceptor which works great with async method calls.  What I'm having problems with now is in my implementation of IAsyncInterceptor, I have injected an IMyAsyncLogger which as the name suggests logs to somewhere asynchronously.  I use this IMyAsyncLogger before and after the invocation.  However, it seems like when calling the IMyAsyncLogger.LogAsync(...), the IAsyncInterceptor implementation seems to be intercepting these calls even though I've not registered it anywhere.  Why is it intercepting the logger calls. Here's my code sample:

Main.cs

```
public static void Main(string[] args)
        {
            RunAsync().GetAwaiter().GetResult();
           
            Console.ReadLine();
        }

public static async Task RunAsync()
        {
            var service = new MyExternalService();

            var proxyGenerator = new ProxyGenerator();
            var interceptor = new MyAsyncInterceptor(new MyAsyncLogger());
            var proxiedExternalService = proxyGenerator.CreateInterfaceProxyWithTarget<IMyExternalService>(service, interceptor);

            await proxiedExternalService.DoStuffAsync().ConfigureAwait(false);

            Console.WriteLine("-----------------");

            var value  = await proxiedExternalService.GetValueAsync().ConfigureAwait(false);

            Console.WriteLine($"My External Service is {value}");
        }
```

MyAsyncInterceptor.cs
```
public class MyAsyncInterceptor : IAsyncInterceptor
    {
        private readonly IMyAsyncLogger _logger;

        public MyAsyncInterceptor(IMyAsyncLogger logger)
        {
            _logger = logger;
        }
        public void InterceptSynchronous(IInvocation invocation)
        {
            invocation.Proceed();
        }

        public void InterceptAsynchronous(IInvocation invocation)
        {
            invocation.ReturnValue = InterceptAsync(invocation);
        }

        public void InterceptAsynchronous<TResult>(IInvocation invocation)
        {
            invocation.ReturnValue = InterceptAsync<TResult>(invocation);
        }

        private async Task InterceptAsync(IInvocation invocation)
        {
            await _logger.LogAsync($"Before Task MyAsyncInterceptor execution").ConfigureAwait(false);
            //Console.WriteLine("Before Task MyAsyncInterceptor execution");

            invocation.Proceed();

            var task = (Task) invocation.ReturnValue;

            await task;

            //Console.WriteLine("After Task MyAsyncInterceptor execution");
            await _logger.LogAsync($"After Task MyAsyncInterceptor execution").ConfigureAwait(false);
        }

        private async Task<TResult> InterceptAsync<TResult>(IInvocation invocation)
        {
            //await _logger.LogAsync($"Before Task<TResult> MyAsyncInterceptor execution").ConfigureAwait(false);
            Console.WriteLine("Before Task<TResult> MyAsyncInterceptor execution");

            invocation.Proceed();
            var task = (Task<TResult>) invocation.ReturnValue;
            TResult result = await task;

            //await _logger.LogAsync($"After Task<TResult> MyAsyncInterceptor execution").ConfigureAwait(false);
            Console.WriteLine("After Task<TResult> MyAsyncInterceptor execution");

            return result;
        }
    }
```

MyAsyncLogger.cs
   ```
 public class MyAsyncLogger : IMyAsyncLogger
    {
        public async Task LogAsync(string message)
        {
            await Task.Delay(2000);
            Console.WriteLine(message);
        }
    }
```

**Comment by JSkimming** [comment]
You've hit an issue we found last year. I experimented with PR #27 but with no luck.

We concluded it's an issue with Castle.Core where it does not expect the interceptor to return before the call to `invocation.Proceed();` which is exactly what happens when you have an asynchronous call before `invocation.Proceed();`

I've commented [here](https://github.com/castleproject/Core/issues/145#issuecomment-365404690) on an issue in Castle.Core castleproject/Core#145, and provided an example PR demonstratoing the ussue castleproject/Core#337.

So, I'm afraid I just haver to say sorry ¯\\_(ツ)_/¯, a number of people have tried to resolve this without succsess.

**Comment by assadfaizi** [comment]
That's a shame. I guess I'll watch out for a fix within Castle.Core hoping someone will know how to do it.  Thanks for the response.

**Comment by xXAvoraXx** [comment]
Is there any solution?

**Comment by JSkimming** [comment]
@xXAvoraXx yes it was fixed by #54 and released with version 2.0.0. See [here](https://github.com/JSkimming/Castle.Core.AsyncInterceptor/releases) for the latest releases.

**Comment by xXAvoraXx** [comment]
> @xXAvoraXx yes it was fixed by #54 and released with version 2.0.0. See [here](https://github.com/JSkimming/Castle.Core.AsyncInterceptor/releases) for the latest releases.

I found this answer for the infinite loop problem. Do you think it's true?

https://github.com/castleproject/Core/blob/master/docs/dynamicproxy-async-interception.md#using-invocationproceed-in-combination-with-asyncawait

**Comment by JSkimming** [comment]
Yes, I think it's true, which is why it was implemented in #54.

The readme on this repo explains the limitations, and if you [avoid option 2](https://github.com/JSkimming/Castle.Core.AsyncInterceptor#option-2-extend-asyncinterceptorbase-class-to-intercept-invocations), it should work.
